src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalescingAssignmentTests.cs

```diff
@@ -0,0 +1,1408 @@
+                //         c ??= new C();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "c").WithArguments("c").WithLocation(7, 9),
+                // (13,9): error CS0165: Use of unassigned local variable 'y1'
+                //         y1.ToString();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y1").WithArguments("y1").WithLocation(13, 9),
+                // (19,9): error CS0165: Use of unassigned local variable 'y2'
+                //         y2.ToString();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y2").WithArguments("y2").WithLocation(19, 9)
+            );
+        }
+
+        [Fact]
+        public void NonStaticInStaticContext()
+        {
+            CreateCompilation(@"
+public class C
+{
+    C P { get; set; }
+    public static void Main()
+    {
+        P ??= new C();
+    }
+}").VerifyDiagnostics(
+                // (7,9): error CS0120: An object reference is required for the non-static field, method, or property 'C.P'
+                //         P ??= new C();
+                Diagnostic(ErrorCode.ERR_ObjectRequired, "P").WithArguments("C.P").WithLocation(7, 9)
+            );
+        }
+
+        [Fact]
+        public void ThrowExpressionRHS()
+        {
+            CompileAndVerify(@"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        try
+        {
+            object o = null;
+            o ??= throw new Exception();
+        }
+        catch (Exception)
+        {
+            Console.WriteLine(""Succeeded"");
+        }
+    }
+}
+", expectedOutput: "Succeeded");
+
+            CreateCompilation(@"
+public class C
+{
+    static object Property { get; }
+    public static void Main()
+    {
+        Property ??= throw null;
+    }
+}
+").VerifyDiagnostics(
+                // (7,9): error CS0200: Property or indexer 'C.Property' cannot be assigned to -- it is read only
+                //         Property ??= throw null;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyProp, "Property").WithArguments("C.Property").WithLocation(7, 9)
+            );
+        }
+
+        [Fact]
+        public void TypeParameterLHS()
+        {
+            CreateCompilation(@"
+class C
+{
+    void M<T>(T t)
+    {
+        t ??= default;
+    }
+}
+").VerifyDiagnostics(
+                // (6,9): error CS0019: Operator '??=' cannot be applied to operands of type 'T' and 'default'
+                //         t ??= default;
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "t ??= default").WithArguments("??=", "T", "default").WithLocation(6, 9)
+            );
+
+            CompileAndVerify(@"
+using System;
+class C
+{
+    static void Main()
+    {
+        Verify<object>(null, ""Assignment Evaluated"");
+    }
+    static void Verify<T>(T t1, T t2) where T : class
+    {
+        Console.WriteLine(t1 ??= t2);
+    }
+}", expectedOutput: "Assignment Evaluated");
+        }
+    }
+}
```

Comment:
Please update `docs/contributing/Compiler%20Test%20Plan.md`

I'd recommending testing semantic model in this PR, since you're adding these tests.
Other test ideas:
- chaining: `x ??= y ??= z`?
- checked vs. unchecked: `nullableIntLocal ??= longValue`

Let's take a note to consider how this feature interacts with nullable feature.

Reply:
I'm going to do semantic model testing in a follow up pr, along with IOperation and CFG work.

> is there a scenario where the LHS is a pointer? Should that work too? If yes, also test void pointer.

`??` doesn't work for pointers, though there is a proposal out to enable this: [https://github.com/dotnet/csharplang/issues/418](https://github.com/dotnet/csharplang/issues/418).

> I assume that precedence in parsing has been addressed in the PR that did the parsing work.

Yes, @gafter had me add a test on that.

As I don't expect the output of the other tests you suggested here to change my implementation, I'll reserve them for a follow up PR.

---
In reply to: [200784195](https://github.com/dotnet/roslyn/pull/28253#discussion_r200784195) [](ancestors = 200784195)